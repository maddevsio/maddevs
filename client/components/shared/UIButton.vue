<template>
  <button
    v-WaveAnimation="!disabled"
    :class="{ 'ui-button--disabled': disabled }"
    class="ui-button"
    @click="handleClick"
  >
    <span v-if="loading">Waiting...</span>
    <slot v-else />
  </button>
</template>

<script>
import WaveAnimation from '@/directives/WaveAnimation'

export default {
  name: 'UIButton',
  directives: {
    WaveAnimation,
  },

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    disabled: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    handleClick() {
      if (!this.disabled) this.$emit('click')
    },
  },
}
</script>

<style lang="scss" scoped>
.ui-button {
  @include font('Inter', 16px, 700);
  border: 1px solid $border-color--red-opacity;
  background-color: $bgcolor--red;
  border-radius: 4px;
  color: $text-color--white;
  letter-spacing: -0.02em;
  line-height: 24px;
  box-shadow: none;
  transition: 0.4s;
  cursor: pointer;

  &:not(.ui-button--disabled) {
    &:hover {
      background-color: $bgcolor--red;
      color: $text-color--white;
    }

    &:active {
      background-color: $button-active--red;
      border-color: $button-active-border--red;
    }
  }

  &--disabled {
    opacity: 0.5;
    cursor: not-allowed;
    user-select: none;
  }

  &--transparent-bgc {
    background-color: transparent;
    color: $text-color--red;
  }

  &_submit-button-footer {
    @include font('Inter', 16px, 400);
    width: 100%;
    height: 54px;
    line-height: 27px;
    letter-spacing: -0.1px;
    font-weight: 400;
  }

  @media only screen and (max-width: 360px) {
    font-size: 14px;
  }
}
</style>
