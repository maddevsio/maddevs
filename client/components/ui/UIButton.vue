<template>
  <button v-WaveAnimation="!disabled" :class="{ 'ui-button--disabled': disabled }" class="ui-button" @click="onClick">
    <span v-if="loading === true">Waiting...</span>
    <slot v-else></slot>
  </button>
</template>

<script>
import WaveAnimation from '@/directives/WaveAnimation'

export default {
  name: 'UIButton',
  directives: {
    WaveAnimation,
  },

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    disabled: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onClick() {
      if (!this.disabled) this.$emit('click')
    },
  },
}
</script>

<style lang="scss" scoped>
@import '../../assets/styles/vars';

.ui-button {
  border: 1px solid $border-color--red-opacity;
  background-color: $bgcolor--red;
  border-radius: 4px;
  color: $text-color--white;
  font-size: 16px;
  font-family: 'Poppins-Bold', sans-serif;
  font-weight: 700;
  letter-spacing: -0.02em;
  line-height: 24px;
  box-shadow: none;
  transition: 0.4s;
  cursor: pointer;

  &:not(.ui-button--disabled) {
    &:hover {
      background-color: $bgcolor--red;
      color: $text-color--white;
    }

    &:active {
      background-color: $button-active--red;
      border-color: $button-active-border--red;
    }
  }

  &--disabled {
    opacity: 0.5;
    cursor: not-allowed;
    user-select: none;
  }

  &--transparent-bgc {
    background-color: transparent;
    color: $text-color--red;
  }

  @media only screen and (max-width: 360px) {
    font-size: 14px;
  }
}
</style>
