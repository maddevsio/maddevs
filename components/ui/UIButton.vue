<template>
  <button class="ui-button" @click="handleClick">{{ text }}</button>
</template>

<script>
export default {
  name: 'UIButton',
  props: {
    text: {
      type: String,
      default: 'Button'
    }
  },
  methods: {
    handleClick() {
      this.$emit('UIButtonClick', true);
    }
  }
};
</script>

<style lang="scss" scoped>
  @import '../../assets/styles/vars';

  .ui-button {
    width: 340px;
    max-height: 70px;
    min-height: 70px;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: $accent-color--red;
    border-radius: 5px;
    color: $text-color--white;
    font-size: 24px;
    font-family: 'MADEEvolveSans-medium', sans-serif;
    font-weight: normal;
    text-decoration: none;
    padding: 20px 40px;
    box-sizing: border-box;
    transition: all .2s ease;
    border: 0;
    cursor: pointer;

    &:hover {
      background-color: darken($accent-color--red, 8);
    }
  }
</style>
