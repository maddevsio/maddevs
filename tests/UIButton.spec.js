import { render, fireEvent, screen } from '@testing-library/vue'
import UIButton from '@/components/ui/UIButton'

describe('uIButton component', () => {
  it('should render correctly with slot', () => {
    const { container } = render(UIButton, {
      slots: {
        default: 'It is a button',
      },
    })

    expect(screen.getByText('It is a button')).not.toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should render correctly without slot', () => {
    const { container } = render(UIButton)

    expect(screen.queryByText('It is a button')).toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should render correctly with loading prop', () => {
    const { container } = render(UIButton, {
      props: {
        loading: true,
      },
    })

    expect(screen.queryByText('Waiting...')).not.toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should correctly handle click', async () => {
    const { emitted } = render(UIButton, {
      slots: {
        default: 'Button',
      },
    })

    const btn = screen.getByText('Button')
    await fireEvent.click(btn)
    expect(emitted().click).toHaveLength(1)
  })

  it('should not handle click if disabled', async () => {
    const { emitted } = render(UIButton, {
      props: {
        disabled: true,
      },
      slots: {
        default: 'Button',
      },
    })

    const btn = screen.getByText('Button')
    await fireEvent.click(btn)
    expect(emitted().click).not.toBeDefined()
  })
})
