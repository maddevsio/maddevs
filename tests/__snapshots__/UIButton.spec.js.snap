// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uIButton component should render correctly with loading prop 1`] = `
<div>
  <button
    class="ui-button"
    style="position: relative; overflow: hidden;"
  >
    <span>
      Waiting...
    </span>
    <div
      class="wave"
    />
  </button>
</div>
`;

exports[`uIButton component should render correctly with slot 1`] = `
<div>
  <button
    class="ui-button"
    style="position: relative; overflow: hidden;"
  >
    It is a button
    <div
      class="wave"
    />
  </button>
</div>
`;

exports[`uIButton component should render correctly without slot 1`] = `
<div>
  <button
    class="ui-button"
    style="position: relative; overflow: hidden;"
  >
    <div
      class="wave"
    />
  </button>
</div>
`;
